File: opticprep/quiz.py
import random
from collections import OrderedDict
from collections.abc import Iterable, Sequence

from opticprep.models import Question, normalize_prompt


def select_questions(
    bank: Sequence[Question],
    count: int,
    topics: Iterable[str],
    difficulties: Iterable[str],
    *,
    rng: random.Random | None = None,
) -> list[Question]:
    topic_set, difficulty_set = set(topics), set(difficulties)
    eligible = [q for q in bank if q.topic in topic_set and q.difficulty in difficulty_set]
    (rng or random).shuffle(eligible)
    return eligible[: min(count, len(eligible))]


def score_questions(questions: Sequence[Question], answers: dict[str, int]) -> dict:
    breakdown: OrderedDict[str, dict[str, int]] = OrderedDict()
    missed: list[Question] = []
    correct = 0
    for question in questions:
        row = breakdown.setdefault(question.topic, {"correct": 0, "total": 0})
        row["total"] += 1
        if answers.get(question.id) == question.correctIndex:
            correct += 1
            row["correct"] += 1
        else:
            missed.append(question)
    return {
        "correct": correct,
        "total": len(questions),
        "by_topic": [{"topic": topic, **values} for topic, values in breakdown.items()],
        "missed": missed,
    }


def merge_unique(
    curated: Sequence[Question], generated: Sequence[Question], limit: int
) -> list[Question]:
    candidates: list[Question] = []
    for index in range(max(len(curated), len(generated))):
        if index < len(generated):
            candidates.append(generated[index])
        if index < len(curated):
            candidates.append(curated[index])
    seen: set[str] = set()
    result: list[Question] = []
    for question in candidates:
        key = normalize_prompt(question.prompt)
        if key and key not in seen:
            seen.add(key)
            result.append(question)
        if len(result) == limit:
            break
    return result

File: lib/quiz.ts
import { curatedQuestions } from "./questions";
import type { HistoryEntry, Question, QuizAnswer, QuizSettings, Topic } from "./types";

export const defaultSettings: QuizSettings = {
  count: 10,
  topics: ["Optics fundamentals", "Ophthalmic lenses", "Frame fitting & measurements", "Dispensing", "Safety & regulatory basics"],
  difficulties: ["foundation", "applied", "challenge"],
  mode: "curated"
};

export function shuffle<T>(items: T[], random = Math.random): T[] {
  const copy = [...items];
  for (let i = copy.length - 1; i > 0; i--) {
    const j = Math.floor(random() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
}

export function selectCurated(settings: QuizSettings, random = Math.random): Question[] {
  const eligible = curatedQuestions.filter(
    (q) => settings.topics.includes(q.topic) && settings.difficulties.includes(q.difficulty)
  );
  return shuffle(eligible, random).slice(0, Math.min(settings.count, eligible.length));
}

export function scoreAnswers(questions: Question[], answers: QuizAnswer[]) {
  const byId = new Map(answers.map((a) => [a.questionId, a]));
  const topicRows = new Map<Topic, { correct: number; total: number }>();
  for (const question of questions) {
    const row = topicRows.get(question.topic) ?? { correct: 0, total: 0 };
    row.total++;
    if (byId.get(question.id)?.correct) row.correct++;
    topicRows.set(question.topic, row);
  }
  return {
    correct: questions.filter((question) => byId.get(question.id)?.correct).length,
    total: questions.length,
    byTopic: [...topicRows.entries()].map(([topic, values]) => ({ topic, ...values })),
    missed: questions.filter((q) => !byId.get(q.id)?.correct)
  };
}

export function makeHistoryEntry(questions: Question[], answers: QuizAnswer[]): HistoryEntry {
  const result = scoreAnswers(questions, answers);
  return {
    id: `${Date.now()}-${Math.random().toString(36).slice(2, 7)}`,
    completedAt: new Date().toISOString(),
    score: result.correct,
    total: result.total,
    topics: [...new Set(questions.map((q) => q.topic))]
  };
}

export function normalizePrompt(value: string) {
  return value.toLocaleLowerCase().replace(/[^\p{L}\p{N}]+/gu, " ").trim();
}

export function mergeUniqueQuestions(curated: Question[], generated: Question[], limit: number) {
  const seen = new Set<string>();
  const interleaved: Question[] = [];
  const maxLength = Math.max(curated.length, generated.length);
  for (let index = 0; index < maxLength; index++) {
    if (curated[index]) interleaved.push(curated[index]);
    if (generated[index]) interleaved.push(generated[index]);
  }
  return interleaved.filter((q) => {
    const key = normalizePrompt(q.prompt);
    if (!key || seen.has(key)) return false;
    seen.add(key);
    return true;
  }).slice(0, limit);
}
